### PR TITLE
Add linter configuration using ESLint, Prettier, and Husky

### LINKED ISSUE #6: Setup ESLint and Prettier for TypeScript
The project needs linting.

We'd like to use ESLint and Prettier to automatically lint typescript as well as any vanilla js files.

Ideally, the linter would prevent git commits on lint fails.

### PR DESCRIPTION
**Pull request by nikoheikkila**
Fixes #6

Added ESLint, Prettier and Husky (with `lint-staged` configuration). Rulesets are based on the aforementioned tool recommendations but I added a couple of exceptions to get this shipped without huge source changes. Let me know if those should be adjusted. I can squash commits in the end.

Run `npm run lint` to check for code style issues, and `npm run fix` to have them fixed automatically. As long as Husky is operating on background, committing style violating changes should be impossible since any errors are fixed automatically.

### CODE REVIEW

**Comment by BrianEmilius** [comment]
It just occurred to me that you might have missed adding the config files explicitly with `git add path/to/file -f`

.gitignore is set to ignore everything, so you have to explicitly add/stage things to the commit, unless they are already tracked.

**Comment by nikoheikkila** [comment]
> It just occurred to me that you might have missed adding the config files explicitly with `git add path/to/file -f`
> 
> .gitignore is set to ignore everything, so you have to explicitly add/stage things to the commit, unless they are already tracked.

Indeed, missed that ignore rule totally. They are now pushed.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+        "prettier/@typescript-eslint", // Uses eslint-config-prettier to disable ESLint rules from @typescript-eslint/eslint-plugin that would conflict with prettier
+        "plugin:prettier/recommended" // Enables eslint-plugin-prettier and eslint-config-prettier. This will display prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
+    ],
+    parserOptions: {
+        ecmaVersion: 2018, // Allows for the parsing of modern ECMAScript features
+        sourceType: "module" // Allows for the use of imports
+    },
+    rules: {
+        "@typescript-eslint/no-var-requires": "off",
+        "@typescript-eslint/no-explicit-any": "off"
```
Comment (BrianEmilius) [inline .eslintrc.js]: Yeah, I think this is necessary too, as not all libraries have types and so far there aren't any custom type definitions in the project.

File: .prettierrc.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    semi: false,
+    singleQuote: false,
+    printWidth: 120,
+    tabWidth: 4,
```
Comment (BrianEmilius) [inline .prettierrc.js]: I strongly believe tabs are better than spaces - it's not a code style choice for me, but an accessibility choice.
Some programmers with impaired vision need the option to scale the tab-wdth up or down so they can better navigate the code at a glance. Tabs gives us this.
However, I am not sure if setting a fixed width of 4 spaces for tabs will prevent this. Do you know? What are your thoughts on this?

File: .prettierrc.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    semi: false,
```
Comment (BrianEmilius) [inline .prettierrc.js]: Who needs semicolons anyway ;)

The very few edge cases where JavaScripts ASI fails are so rare that they're not even worth discussing in my opinion.

File: .eslintrc.js (hunk 1 above)
Comment (nikoheikkila) [reply .eslintrc.js]: Agree. This could also be a `"warn"` so e.g. CI won't fail on it but the developer is notified.

File: .prettierrc.js (hunk 2 above)
Comment (nikoheikkila) [reply .prettierrc.js]: There are definitely two schools of thought on this matter. Personally, I go with either way as I usually include EditorConfig in my projects which will automatically set my editor settings to comply with project.

Hard tabs or spaces with a width of 4 is usually a good choice unless the logic includes a lot of nesting (which is usually a code smell, of course).

File: .prettierrc.js (hunk 3 above)
Comment (nikoheikkila) [reply .prettierrc.js]: Yeah, semi-colons make code unaesthetic in my eyes anyway. 🖌

**Comment by nikoheikkila** [comment]
> Is husky missing? Or did I miss something?
> 
> I made a few in-line comments.

Husky configuration is in `package.json`. It also supports its own configuration file if that is preferred.

**Comment by BrianEmilius** [comment]
Ah, that's my bad then!
This is a top-quality contribution. Thank you so much for working on it!

I am going to merge. Would you like a mention in the "contributors" file?

**Comment by nikoheikkila** [comment]
> Ah, that's my bad then!
> This is a top-quality contribution. Thank you so much for working on it!
> 
> I am going to merge. Would you like a mention in the "contributors" file?

Sure, please add me to contributors. I could take a look at #7 next since TypeDoc is a new thing for me. I won't _reserve_ it yet if someone else wants to grab it. 😄

**Comment by BrianEmilius** [comment]
Sounds great! Go ahead and have a swing at it. I'll add your name to the list of contribs :smile:

**Pull request merged**